// zarrita/dist/src/indexing/ops.js
import { get as get_with_setter } from "./get.js";
import { set as set_with_setter } from "./set.js";
/** A 1D "view" of an array that can be used to set values in the array. */
function objectArrayView(arr, offset = 0, size) {
    let length = size ?? arr.length - offset;
    return {
        length,
        subarray(from, to = length) {
            return objectArrayView(arr, offset + from, to - from);
        },
        set(data, start = 0) {
            for (let i = 0; i < data.length; i++) {
                arr[offset + start + i] = data.get(i);
            }
        },
        get(index) {
            return arr[offset + index];
        },
    };
}
/**
 * Convert a chunk to a Uint8Array that can be used with the binary
 * set functions. This is necessary because the binary set functions
 * require a contiguous block of memory, and allows us to support more than
 * just the browser's TypedArray objects.
 *
 * WARNING: This function is not meant to be used directly and is NOT type-safe.
 * In the case of `Array` instances, it will return a `objectArrayView` of
 * the underlying, which is supported by our binary set functions.
 */
function compatChunk(arr) {
    if (globalThis.Array.isArray(arr.data)) {
        return {
            // @ts-expect-error
            data: objectArrayView(arr.data),
            stride: arr.stride,
            bytesPerElement: 1,
        };
    }
    return {
        data: new Uint8Array(arr.data.buffer, arr.data.byteOffset, arr.data.byteLength),
        stride: arr.stride,
        bytesPerElement: arr.data.BYTES_PER_ELEMENT,
    };
}
/** Hack to get the constructor of a typed array constructor from an existing TypedArray. */
function getTypedArrayConstructor(arr) {
    if ("chars" in arr) {
        // our custom TypedArray needs to bind the number of characters per
        // element to the constructor.
        return arr.constructor.bind(null, arr.chars);
    }
    return arr.constructor;
}
/**
 * Convert a scalar to a Uint8Array that can be used with the binary
 * set functions. This is necessary because the binary set functions
 * require a contiguous block of memory, and allows us to support more
 * than just the browser's TypedArray objects.
 *
 * WARNING: This function is not meant to be used directly and is NOT type-safe.
 * In the case of `Array` instances, it will return a `objectArrayView` of
 * the scalar, which is supported by our binary set functions.
 */
function compatScalar(arr, value) {
    if (globalThis.Array.isArray(arr.data)) {
        // @ts-expect-error
        return objectArrayView([value]);
    }
    let TypedArray = getTypedArrayConstructor(arr.data);
    // @ts-expect-error - value is a scalar and matches
    let data = new TypedArray([value]);
    return new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
}
export const setter = {
    prepare(data, shape, stride) {
        return { data, shape, stride };
    },
    setScalar(dest, sel, value) {
        let view = compatChunk(dest);
        setScalarBinary(view, sel, compatScalar(dest, value), view.bytesPerElement);
    },
    setFromChunk(dest, src, projections) {
        let view = compatChunk(dest);
        setFromChunkBinary(view, compatChunk(src), view.bytesPerElement, projections);
    },
};
/** @category Utility */
export async function get(arr, selection = null, opts = {}) {
    return get_with_setter(arr, selection, opts, setter);
}
/** @category Utility */
export async function set(arr, selection, value, opts = {}) {
    return set_with_setter(arr, selection, value, opts, setter);
}
function indicesLen(start, stop, step) {
    if (step < 0 && stop < start) {
        return Math.floor((start - stop - 1) / -step) + 1;
    }
    if (start < stop)
        return Math.floor((stop - start - 1) / step) + 1;
    return 0;
}
function setScalarBinary(out, outSelection, value, bytesPerElement) {
    if (outSelection.length === 0) {
        out.data.set(value, 0);
        return;
    }
    const [slice, ...slices] = outSelection;
    const [currStride, ...stride] = out.stride;
    if (typeof slice === "number") {
        const data = out.data.subarray(currStride * slice * bytesPerElement);
        setScalarBinary({ data, stride }, slices, value, bytesPerElement);
        return;
    }
    const [from, to, step] = slice;
    const len = indicesLen(from, to, step);
    if (slices.length === 0) {
        for (let i = 0; i < len; i++) {
            out.data.set(value, currStride * (from + step * i) * bytesPerElement);
        }
        return;
    }
    for (let i = 0; i < len; i++) {
        const data = out.data.subarray(currStride * (from + step * i) * bytesPerElement);
        setScalarBinary({ data, stride }, slices, value, bytesPerElement);
    }
}
function setFromChunkBinary(dest, src, bytesPerElement, projections) {
    const [proj, ...projs] = projections;
    const [dstride, ...dstrides] = dest.stride;
    const [sstride, ...sstrides] = src.stride;
    if (proj.from === null) {
        if (projs.length === 0) {
            dest.data.set(src.data.subarray(0, bytesPerElement), proj.to * bytesPerElement);
            return;
        }
        setFromChunkBinary({
            data: dest.data.subarray(dstride * proj.to * bytesPerElement),
            stride: dstrides,
        }, src, bytesPerElement, projs);
        return;
    }
    if (proj.to === null) {
        if (projs.length === 0) {
            let offset = proj.from * bytesPerElement;
            dest.data.set(src.data.subarray(offset, offset + bytesPerElement), 0);
            return;
        }
        setFromChunkBinary(dest, {
            data: src.data.subarray(sstride * proj.from * bytesPerElement),
            stride: sstrides,
        }, bytesPerElement, projs);
        return;
    }
    const [from, to, step] = proj.to;
    const [sfrom, _, sstep] = proj.from;
    const len = indicesLen(from, to, step);
    if (projs.length === 0) {
        // NB: we have a contiguous block of memory
        // so we can just copy over all the data at once.
        if (step === 1 && sstep === 1 && dstride === 1 && sstride === 1) {
            let offset = sfrom * bytesPerElement;
            let size = len * bytesPerElement;
            dest.data.set(src.data.subarray(offset, offset + size), from * bytesPerElement);
            return;
        }
        // Otherwise, we have to copy over each element individually.
        for (let i = 0; i < len; i++) {
            let offset = sstride * (sfrom + sstep * i) * bytesPerElement;
            dest.data.set(src.data.subarray(offset, offset + bytesPerElement), dstride * (from + step * i) * bytesPerElement);
        }
        return;
    }
    for (let i = 0; i < len; i++) {
        setFromChunkBinary({
            data: dest.data.subarray(dstride * (from + i * step) * bytesPerElement),
            stride: dstrides,
        }, {
            data: src.data.subarray(sstride * (sfrom + i * sstep) * bytesPerElement),
            stride: sstrides,
        }, bytesPerElement, projs);
    }
}
