// @blueprintjs/select/lib/cjs/common/classes.js
"use strict";
/*
 * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUGGEST_POPOVER = exports.SELECT_POPOVER = exports.OMNIBAR_OVERLAY = exports.OMNIBAR = exports.MULTISELECT_TAG_INPUT_INPUT = exports.MULTISELECT_POPOVER_TAG_INPUT_MARGIN = exports.MULTISELECT_POPOVER_DEFAULT_WIDTH = exports.MULTISELECT_POPOVER = exports.MULTISELECT = void 0;
var core_1 = require("@blueprintjs/core");
var NS = core_1.Classes.getClassNamespace();
exports.MULTISELECT = "".concat(NS, "-multi-select");
exports.MULTISELECT_POPOVER = "".concat(exports.MULTISELECT, "-popover");
exports.MULTISELECT_POPOVER_DEFAULT_WIDTH = "".concat(exports.MULTISELECT, "-popover-default-width");
exports.MULTISELECT_POPOVER_TAG_INPUT_MARGIN = "".concat(exports.MULTISELECT, "-popover-tag-input-margin");
exports.MULTISELECT_TAG_INPUT_INPUT = "".concat(exports.MULTISELECT, "-tag-input-input");
exports.OMNIBAR = "".concat(NS, "-omnibar");
exports.OMNIBAR_OVERLAY = "".concat(exports.OMNIBAR, "-overlay");
exports.SELECT_POPOVER = "".concat(NS, "-select-popover");
exports.SUGGEST_POPOVER = "".concat(NS, "-suggest-popover");
