// @blueprintjs/core/lib/cjs/common/index.js
"use strict";
/*
 * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = exports.Classes = exports.Size = exports.setRef = exports.refHandler = exports.mergeRefs = exports.isRefObject = exports.isRefCallback = exports.getRef = exports.DISPLAYNAME_PREFIX = exports.removeNonHTMLProps = exports.Position = exports.Keys = exports.Intent = exports.Elevation = exports.ButtonVariant = exports.Boundary = exports.TextAlignment = exports.Alignment = exports.AbstractPureComponent = exports.AbstractComponent = exports.Colors = void 0;
var tslib_1 = require("tslib");
var colors_1 = require("@blueprintjs/colors");
Object.defineProperty(exports, "Colors", { enumerable: true, get: function () { return colors_1.Colors; } });
var abstractComponent_1 = require("./abstractComponent");
Object.defineProperty(exports, "AbstractComponent", { enumerable: true, get: function () { return abstractComponent_1.AbstractComponent; } });
var abstractPureComponent_1 = require("./abstractPureComponent");
Object.defineProperty(exports, "AbstractPureComponent", { enumerable: true, get: function () { return abstractPureComponent_1.AbstractPureComponent; } });
var alignment_1 = require("./alignment");
Object.defineProperty(exports, "Alignment", { enumerable: true, get: function () { return alignment_1.Alignment; } });
Object.defineProperty(exports, "TextAlignment", { enumerable: true, get: function () { return alignment_1.TextAlignment; } });
var boundary_1 = require("./boundary");
Object.defineProperty(exports, "Boundary", { enumerable: true, get: function () { return boundary_1.Boundary; } });
var buttonVariant_1 = require("./buttonVariant");
Object.defineProperty(exports, "ButtonVariant", { enumerable: true, get: function () { return buttonVariant_1.ButtonVariant; } });
var elevation_1 = require("./elevation");
Object.defineProperty(exports, "Elevation", { enumerable: true, get: function () { return elevation_1.Elevation; } });
var intent_1 = require("./intent");
Object.defineProperty(exports, "Intent", { enumerable: true, get: function () { return intent_1.Intent; } });
var keyCodes_1 = require("./keyCodes");
Object.defineProperty(exports, "Keys", { enumerable: true, get: function () { return keyCodes_1.KeyCodes; } });
var position_1 = require("./position");
Object.defineProperty(exports, "Position", { enumerable: true, get: function () { return position_1.Position; } });
var props_1 = require("./props");
Object.defineProperty(exports, "removeNonHTMLProps", { enumerable: true, get: function () { return props_1.removeNonHTMLProps; } });
Object.defineProperty(exports, "DISPLAYNAME_PREFIX", { enumerable: true, get: function () { return props_1.DISPLAYNAME_PREFIX; } });
var refs_1 = require("./refs");
Object.defineProperty(exports, "getRef", { enumerable: true, get: function () { return refs_1.getRef; } });
Object.defineProperty(exports, "isRefCallback", { enumerable: true, get: function () { return refs_1.isRefCallback; } });
Object.defineProperty(exports, "isRefObject", { enumerable: true, get: function () { return refs_1.isRefObject; } });
Object.defineProperty(exports, "mergeRefs", { enumerable: true, get: function () { return refs_1.mergeRefs; } });
Object.defineProperty(exports, "refHandler", { enumerable: true, get: function () { return refs_1.refHandler; } });
Object.defineProperty(exports, "setRef", { enumerable: true, get: function () { return refs_1.setRef; } });
var size_1 = require("./size");
Object.defineProperty(exports, "Size", { enumerable: true, get: function () { return size_1.Size; } });
var Classes = tslib_1.__importStar(require("./classes"));
exports.Classes = Classes;
var Utils = tslib_1.__importStar(require("./utils"));
exports.Utils = Utils;
// NOTE: Errors is not exported in public API
