// @blueprintjs/core/lib/cjs/context/index.js
"use strict";
/*
 * Copyright 2021 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PortalProvider = exports.PortalContext = exports.OverlaysProvider = exports.OverlaysContext = exports.HotkeysProvider = exports.HotkeysContext = exports.BlueprintProvider = void 0;
var blueprintProvider_1 = require("./blueprintProvider");
Object.defineProperty(exports, "BlueprintProvider", { enumerable: true, get: function () { return blueprintProvider_1.BlueprintProvider; } });
var hotkeysProvider_1 = require("./hotkeys/hotkeysProvider");
Object.defineProperty(exports, "HotkeysContext", { enumerable: true, get: function () { return hotkeysProvider_1.HotkeysContext; } });
Object.defineProperty(exports, "HotkeysProvider", { enumerable: true, get: function () { return hotkeysProvider_1.HotkeysProvider; } });
var overlaysProvider_1 = require("./overlays/overlaysProvider");
Object.defineProperty(exports, "OverlaysContext", { enumerable: true, get: function () { return overlaysProvider_1.OverlaysContext; } });
Object.defineProperty(exports, "OverlaysProvider", { enumerable: true, get: function () { return overlaysProvider_1.OverlaysProvider; } });
var portalProvider_1 = require("./portal/portalProvider");
Object.defineProperty(exports, "PortalContext", { enumerable: true, get: function () { return portalProvider_1.PortalContext; } });
Object.defineProperty(exports, "PortalProvider", { enumerable: true, get: function () { return portalProvider_1.PortalProvider; } });
