// @blueprintjs/core/lib/esnext/components/dialog/dialog.js
/*
 * Copyright 2015 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import classNames from "classnames";
import * as React from "react";
import { IconSize, SmallCross } from "@blueprintjs/icons";
import { AbstractPureComponent, Classes, DISPLAYNAME_PREFIX, mergeRefs, } from "../../common";
import * as Errors from "../../common/errors";
import { uniqueId } from "../../common/utils";
import { Button } from "../button/buttons";
import { H6 } from "../html/html";
import { Icon } from "../icon/icon";
import { Overlay2, OVERLAY2_DEFAULT_PROPS } from "../overlay2/overlay2";
/**
 * Dialog component.
 *
 * @see https://blueprintjs.com/docs/#core/components/dialog
 */
export class Dialog extends AbstractPureComponent {
    static defaultProps = {
        canOutsideClickClose: true,
        isOpen: false,
    };
    childRef = React.createRef();
    titleId;
    static displayName = `${DISPLAYNAME_PREFIX}.Dialog`;
    constructor(props) {
        super(props);
        const id = uniqueId("bp-dialog");
        this.titleId = `title-${id}`;
    }
    render() {
        const { className, children, containerRef, style, title, role = "dialog", ...overlayProps } = this.props;
        return (React.createElement(Overlay2, { ...overlayProps, className: Classes.OVERLAY_SCROLL_CONTAINER, childRef: this.childRef, hasBackdrop: true },
            React.createElement("div", { className: Classes.DIALOG_CONTAINER, ref: mergeRefs(containerRef, this.childRef) },
                React.createElement("div", { className: classNames(Classes.DIALOG, className), role: role, "aria-modal": overlayProps.enforceFocus ?? OVERLAY2_DEFAULT_PROPS.enforceFocus, "aria-labelledby": this.props["aria-labelledby"] || (title ? this.titleId : undefined), "aria-describedby": this.props["aria-describedby"], style: style },
                    this.maybeRenderHeader(),
                    children))));
    }
    validateProps(props) {
        if (props.title == null) {
            if (props.icon != null) {
                console.warn(Errors.DIALOG_WARN_NO_HEADER_ICON);
            }
            if (props.isCloseButtonShown != null) {
                console.warn(Errors.DIALOG_WARN_NO_HEADER_CLOSE_BUTTON);
            }
        }
    }
    maybeRenderCloseButton() {
        // show close button if prop is undefined or null
        // this gives us a behavior as if the default value were `true`
        if (this.props.isCloseButtonShown !== false) {
            return (React.createElement(Button, { "aria-label": "Close", className: Classes.DIALOG_CLOSE_BUTTON, icon: React.createElement(SmallCross, { size: IconSize.STANDARD }), onClick: this.props.onClose, variant: "minimal" }));
        }
        else {
            return undefined;
        }
    }
    maybeRenderHeader() {
        const { icon, title } = this.props;
        if (title == null) {
            return undefined;
        }
        return (React.createElement("div", { className: Classes.DIALOG_HEADER },
            React.createElement(Icon, { icon: icon, size: IconSize.STANDARD, "aria-hidden": true, tabIndex: -1 }),
            React.createElement(H6, { id: this.titleId }, title),
            this.maybeRenderCloseButton()));
    }
}
